File: Search a 2D Matrix.py
# Link: https://leetcode.com/problems/search-a-2d-matrix/submissions/

class Solution(object):
    
    def binarySearch(self, matrix, low, high, target):
        
        if high >= low:
            
            # Get midpoint
            midPoint = (high + low) // 2
            
            # If element is at midPoint, return true
            if matrix[midPoint] == target:
                return True
            
            elif matrix[midPoint] > target:
                return self.binarySearch(matrix, low, midPoint - 1, target)
            
            else:
                return self.binarySearch(matrix, midPoint + 1, high, target)
            
        else:
            return False
    
    def searchMatrix(self, matrix, target):
        """
        :type matrix: List[List[int]]
        :type target: int
        :rtype: bool
        """
        
        # If matrix is NoneType, exit function
        if matrix is None:
            return False
        
        # Get midpoint of matrix
        midRow = len(matrix) // 2
        
        # If matrix only has 1 row, continue to while-loop
        if len(matrix) != 1:
            
            # Check if 1st element of midRow equals to target
            if target > matrix[midRow][0]:
                return self.searchMatrix(matrix[midRow:], target)

            elif target < matrix[midRow][0]:
                return self.searchMatrix(matrix[:midRow], target)
                
        # If target is potentially inside midRow, break down row
        # NOTE: Only send a single row from the matrix. Not the matrix!
        return self.binarySearch(matrix[midRow], 0, len(matrix[0]) - 1, target)

File: Search a 2D Matrix.java
// https://leetcode.com/problems/search-a-2d-matrix/
/*
Time Complexity: O(log(m * n)) where m and n are the dimensions of the matrix. 
    Binary search is used to first look for the target row, and then performed again to look for target in said row.

Space Complexity: O(1)
*/

class Solution {
    public boolean searchMatrix(int[][] matrix, int target) {
        int maxRows = matrix.length, maxCols = matrix[0].length;

        // Look for row that potentially has target
        int topPtr = 0, bottomPtr = maxRows - 1;
        while (topPtr <= bottomPtr) {
            int midPtr = topPtr + ((bottomPtr - topPtr) / 2);
            int firstElem = matrix[midPtr][0];
            int lastElem = matrix[midPtr][maxCols - 1];

            if (firstElem > target) { // If first elem > target, then search top-half of matrix
                bottomPtr = midPtr - 1;
            } else if (lastElem < target) { // If last elem < target, then search bottom-half of matrix
                topPtr = midPtr + 1;
            } else { // If first element <= target <= last element in curRow
                break;
            }
        }

        // Check if above condition was broken. If so, target doesn't exist in matrix
        if (topPtr > bottomPtr) {
            return false;
        }

        // Perform binary search again, but on target row
        int leftPtr = 0, rightPtr = maxCols - 1;
        int targetRow = topPtr + ((bottomPtr - topPtr) / 2);
        while (leftPtr <= rightPtr) {

            int midPtr = leftPtr + ((rightPtr - leftPtr) / 2);
            
            if (matrix[targetRow][midPtr] > target) { // If middle element > target, search left-hand side 
                rightPtr = midPtr - 1;
            } else if (matrix[targetRow][midPtr] < target) { // If middle element < target, search right-hand side
                leftPtr = midPtr + 1;
            } else {
                return true;
            }
        }

        return false;
    }
}
